pkg/pgmodel/querier/query_builder.go

```diff
@@ -216,68 +218,107 @@ type aggregators struct {
+	//
+	// This forms the basis for how we determine if we can do a promQL
+	// pushdown: If the current node is a vector selector, we look up the AST
+	// to check if the "grandparent" node to the vector selector is a known
+	// function which we can push down to the database.
+
+	path := metadata.path
+	queryHints := metadata.queryHints
+	selectHints := metadata.selectHints
+
+	vs, isVectorSelector := queryHints.CurrentNode.(*parser.VectorSelector)
 	if isVectorSelector {
-		/* Try to optimize the aggregation first since that would return less data than a plain vector selector */
 		if len(path) >= 2 {
-			//switch on the 2nd-to-last last path node
-			node := path[len(path)-2]
-			callNode, isCall := node.(*parser.Call)
-			if isCall {
-				switch callNode.Func.Name {
-				case "delta":
-					agg, err := callAggregator(hints, callNode.Func.Name)
-					return agg, node, err
-				case "rate", "increase":
-					if rateIncreaseExtensionRange(extension.PromscaleExtensionVersion) {
-						agg, err := callAggregator(hints, callNode.Func.Name)
-						return agg, node, err
-					}
-				}
+			grandparent := path[len(path)-2]
+			funcName, canPushDown := tryExtractPushdownableFunctionName(grandparent)
+			if canPushDown {
+				agg, err := callAggregator(selectHints, funcName)
+				return agg, grandparent, err
 			}
 		}
 
 		//TODO: handle the instant query (hints.Step==0) case too.
-		/* vector selector pushdown improves performance by selecting from the database only the last point
-		* in a vector selector window(step) this decreases the amount of samples transferred from the DB to Promscale
-		* by orders of magnitude. A vector selector aggregate also does not require ordered inputs which saves
-		* a sort and allows for parallel evaluation. */
-		if hints.Step > 0 &&
-			hints.Range == 0 && /* So this is not an aggregate. That's optimized above */
+
+		// vector selector pushdown improves performance by selecting from the
+		// database only the last point in a vector selector window(step).
+		// This decreases the number of samples transferred from the DB to
+		// Promscale by orders of magnitude. A vector selector aggregate also
+		// does not require ordered inputs which saves a sort and allows for
+		// parallel evaluation.
+		if selectHints.Step > 0 &&
+			selectHints.Range == 0 && /* So this is not an aggregate. That's optimized above */
 			!calledByTimestamp(path) &&
 			vs.OriginalOffset == time.Duration(0) &&
 			vs.Offset == time.Duration(0) &&
 			vectorSelectorExtensionRange(extension.PromscaleExtensionVersion) {
 			qf := aggregators{
-				valueClause: "vector_selector($%d, $%d,$%d, $%d, time, value)",
-				valueParams: []interface{}{qh.StartTime, qh.EndTime, hints.Step, qh.Lookback.Milliseconds()},
+				valueClause: "vector_selector($%d, $%d, $%d, $%d, time, value)",
+				valueParams: []interface{}{queryHints.StartTime, queryHints.EndTime, selectHints.Step, queryHints.Lookback.Milliseconds()},
```

Comment:
Thanks for doing this. 

While short names can be beneficial in shorter and simpler functions, here it hurts readability IMO.

Reply:
👍